pkg/storage/ingest/reader.go

```diff
@@ -105,38 +112,109 @@ func (r *PartitionReader) start(ctx context.Context) error {
 	if err != nil {
 		return errors.Wrap(err, "starting service manager")
 	}
+
+	// Enforce the max consumer lag (if enabled).
+	if maxLag := r.kafkaCfg.MaxConsumerLagAtStartup; maxLag > 0 {
+		if err := r.processNextFetchesUntilMaxLagHonored(ctx, maxLag); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
 func (r *PartitionReader) stop(error) error {
 	level.Info(r.logger).Log("msg", "stopping partition reader")
 
-	err := services.StopManagerAndAwaitStopped(context.Background(), r.dependencies)
-	if err != nil {
-		return errors.Wrap(err, "stopping service manager")
+	return r.stopDependencies()
+}
+
+func (r *PartitionReader) stopDependencies() error {
+	if r.dependencies != nil {
+		if err := services.StopManagerAndAwaitStopped(context.Background(), r.dependencies); err != nil {
+			return errors.Wrap(err, "stopping service manager")
+		}
 	}
-	r.client.Close()
+
+	if r.client != nil {
+		r.client.Close()
+	}
+
 	return nil
 }
 
 func (r *PartitionReader) run(ctx context.Context) error {
 	for ctx.Err() == nil {
-		fetches := r.client.PollFetches(ctx)
```

Comment:
Note to reviewers: this block of code as been copied, as is, to `processNextFetches()`.